### ISSUE TITLE
Authentications do not exists yet

### ISSUE THREAD

**Issue by ryan-bradford** [ryan-bradford/WHAM#1]
Right now, every URL request accepts an authentication token, but none actually use it. We need to make it so transactions are actually verified.

**Comment by dajinchu** [comment]
I'm not sure if this is exactly what we want but it could help keep things clean when it comes to authentication: 
http://www.passportjs.org/

**Comment by ryan-bradford** [comment]
I can implement this, but it doesn't solve 100% of our problems. 

We have authentication all set right now (it might not be the cleanest way, but it works). That's the process where the users posts to either /login/patient or /login/therapist to authenticate that they are the person they say they are. That is totally functional right now. 

What is not functional is authorization. When the user queries to any url we need some way of verifying  (even though they are logged in and ARE the person they say they are) that they have the permissions to do what they are trying to do.

In our application, this is especially complicated because (for example) a therapist assigned to a user through the PATIENT_THERAPIST table should be able to send a patient a message, but a therapist not assigned should not be able to. Because of this, it's not as simple as checking if patientID matches the requested patientID (or other simple, direct string comparison stuff). Also, we need to build in some way of not being able to fake the patientID.

This is why I thought passing some randomly-generated, unique user token (I decided to go with the salt) would be a good idea. It's hard/impossible to fake this token, it's totally unique to WHAM, has nothing to do with the user, and can be used to get user info to check if they have proper permissions.

So in summary, passport looks good, but it does not solve 100% of our problems. Also, I do bet there is a library for JS that solves authorization problems (maybe passport does, and I just missed something). Also sorry if I'm just babbling on about stuff you already know...

**Comment by ryan-bradford** [comment]
Check out this library for authorizations: https://github.com/OptimalBits/node_acl.

**Comment by dajinchu** [comment]
Yeah it does appear passport is more authentication than authorization. The token idea sounds good. From the token we can derive exactly who is sending the request. We just need to treat the token with as much caution as their password because it will basically be the key to all of a user's permissions.  I think we should also put the token in a cookie in the browser and that'll handle the login process there. 

One thing is that for security purposes we might want the token to auto expire after a set amount of time, and after that they will have to login again, or if we have some security incident we can refresh all tokens at will to force everyone to login again. Basically if someone's token is stolen, we just expire it and make a new token when they login with their password, rather than forcing the user to make a new password. For that reason I think we might want the token to be a different random hash, not the salt.

**Comment by ryan-bradford** [comment]
That's definitely a good point. I'm changing the login functionality to use passport right now, and then I'll look into a token-generation library and combine it with ACL.

**Comment by dajinchu** [comment]
Awesome. Good summary. This should definitely work. We can look into any security vulnerabilities later I think.
